Add syntax highlighting for Roc docs

lukewilliamboswell opened the pull request:
Adds syntax highlighting for the documentation `roc docs`. Removes some unused code. Refactors the html `highlight` into a crate that is common to `static-site` example platform and the `roc_docs` crate.

Note I haven't completed a `cargo test` run yet, but have manually tested this doesn't break website docs, or tutorial.

`.css` styles for code samples copied from tutorial.

<img width="928" alt="Screen Shot 2023-03-06 at 22 44 45" src="https://user-images.githubusercontent.com/2679227/223101317-9acfba44-6d55-4d64-b279-ec798744afae.png">

lukewilliamboswell commented:
@Anton-4 updated this PR to support fenced code as discussed on [zulip](https://roc.zulipchat.com/#narrow/stream/304641-ideas/topic/code.20in.20docs/near/340052655). I left the actual checking and REPL with TODO's for now. I also updated all of the Tutorial markdown to the new style and manually checked each example. There are a lot of examples that need to be touched up to make more semantic sense, but that is out of scope for this PR.

lukewilliamboswell commented on crates/compiler/builtins/roc/Num.roc:
```diff
@@ -1390,7 +1392,7 @@ toF64Checked : Num * -> Result F64 [OutOfBounds]
 # ## digits, and then zeroes will be added afterwards until it reaches `min` digits.
 # ##
-# ## >>> Num.format 1.23 { decPlaces: 0, decPointVis: AlwaysShow }
+# ##     Num.format 1.23 { decPlaces: 0, decPointVis: AlwaysShow }
```
This while section was weird. It was like this feom a previous PR. I'm not sure but it looks like we've commented all this section out for some reason? Should we just remove the code, or do we have plans to add it back in?

Anton-4 replied:
Oh right it's just commented out, let's remove it, there's no TODO and we'll know where to find it if we want it back.

lukewilliamboswell replied:
I removed all of the commented code.

lukewilliamboswell commented:
Updated to improve token groups as discussed on [zulip](https://roc.zulipchat.com/#narrow/stream/304641-ideas/topic/code.20in.20docs/near/340394575). I experimented with different variations, and this is what I like the most.

I also added a colour scheme which looks nice for both light and dark modes. It uses five colours;

- Cyan for Number, String, Tag, Type literals
- Magenta for Keywords and punctuation
- Orange for Operators
- Violet for Variables modules and field names
- Green for Types, Tags, and Modules

## Light Example

<img width="871" alt="Screen Shot 2023-03-10 at 11 39 49" src="https://user-images.githubusercontent.com/2679227/224193602-b139f90b-058d-4206-acf4-1b8a37fc529c.png">

## Dark Example

<img width="873" alt="Screen Shot 2023-03-10 at 11 39 59" src="https://user-images.githubusercontent.com/2679227/224193612-37b84ac1-d7df-4a70-8193-79db6cf8b341.png">

rtfeldman commented:
Looks cool!

One thing I should have mentioned earlier: since these are docs for everyone (as opposed to an individual's personal editor preference), it's important that we take accessibility into account. I'm specifically thinking contrast ratios, which at a glance look probably too low to pass WCAG AA accessibility conformance at the moment.

When I picked the theme colors for the tutorial and docs I used [this tool](https://accessible-colors.com/)) to get the colors conformant, and it was very helpful!

(There's [some debate](https://uxmovement.com/buttons/the-myths-of-color-contrast-accessibility/) as to whether WCAG AA is optimal for darker text on colored backgrounds in particular, but that probably won't come up here.)

lukewilliamboswell commented:
@rtfeldman I tweaked the colours they should all now pass WCAG AA compliant for both Light and Dark modes. Note I only checked the code snippets and have left the other colours as they were. I tried to clean up some of the unused css along the way, and unify more of the styles between docs and tutorial. There is still more work there to do, but that it out of scope for this PR.

Anton-4 commented on www/generate_tutorial/src/input/tutorial.md:
````diff
@@ -161,7 +161,7 @@ main =
 ```
 
-Now run `roc dev` again. This time the "Downloading …" message won't appear; the file has been cached from last time, and won't need to be downloaded again.
+Now run `roc dev` again. This time the "Downloading ..." message won't appear; the file has been cached from last time, and won't need to be downloaded again.
````
Did the ellipsis symbol cause any issues?

lukewilliamboswell replied:
Yes, it crashes the parser from memory. Was going to put an issue in to track it.

Anton-4 commented:
Thank you very much @lukewilliamboswell! This was a big effort :muscle:

rtfeldman commented:
@lukewilliamboswell hm, it looks like the colors on the code snippets don't pass WCAG AA 😅 

e.g. here's `getRatings`

<img width="413" alt="Screen Shot 2023-0311 at 5 00 04 PM" src="https://user-images.githubusercontent.com/1094080/224513217-e6adf6d1-56c7-4d7d-8e26-ffd0c0a226f5.png">

<img width="936" alt="Screen Shot 2023-03-11 at 4 59 12 PM" src="https://user-images.githubusercontent.com/1094080/224513224-158b0024-5746-4c2e-93b5-6b5fedd1a55f.png">

Would you mind taking a second pass at the various foreground/background combos in light and dark mode with [accessible-colors.com](accessible-colors.com)? It has some quick fix suggestions that I've found helpful in the past, e.g. changing the background color to #111423 might address more than one contrast issue here. 😄

rtfeldman commented:
Actually, darkening the background that much makes it kinda hard to see that code snippets have a different background at all...

<img width="865" alt="Screen Shot 2023-03-11 at 5 08 43 PM" src="https://user-images.githubusercontent.com/1094080/224513548-eb9fd81c-b044-40eb-a969-3592cd4f3d58.png">

...so maybe a combination of lightening the text a bit and darkening the background a bit? I often try these tweaks by changing from RGB to HSL representation and then changing just L to make it lighter or darker. (Can always change it back to a hex representation afterwards if desired.)
